[클린코드 리액트 3기 박재영] 계산기 구현

SaeWooKKang opened the pull request:
안녕하세요 우디님! 미션 완료하여 리뷰 부탁드리겠습니다! 🙇‍♂️

제가 구현하면서 생각했던 내용을 적으면 리뷰 하실 때 도움이 될까 싶어서 과정을 자세하게 작성했고, 궁금한 사항도 작성했습니다. 추가로 구현 과정 혹은 결과에서 개선할 사항이 있다면 말씀해주시면 감사하겠습니다!

## 구현

### [feat(Calculator/interface.ts): interface 작성](https://github.com/next-step/react-calculator/commit/f68cc38ebdb4850a1d8e14a998966da823f79b90)
계산기 미션에서 `계산 로직`은 `ui` 혹은 더 나아가 `라이브러리`가 달라져도 재사용할 수 있다고 생각했습니다. 그래서 해당 **로직을 분리**(moduels/Calculator)하여 작성했습니다. 

먼저 `Calculator`에 필요한 `Interface`를 작성했습니다. 필요한 메서드는 다르지만 `함수 시그니처`는 동일하여 타입을 선언해 **재사용** 했습니다.

### [feat(Calculator.ts): interface 구현](https://github.com/next-step/react-calculator/commit/842d298ac4fe61a249b24f65c791e7cb97adda2e)

그 다음 `interface`를 구현 했습니다. 
다른 연산들과 달리 나누기 연산에서 '함수 파라미터로 자릿수를 받을수 있도록 구현하는게 좋을까?' 에 대해서 고민이 됐습니다.
제 생각에는 '사용하는 쪽에서 결과값을 가공해서 보여주면 되지 않을까?' 싶어서 단순히 연산만 처리하도록 작성했습니다. 우디님이라면 어떻게 처리 하실지 궁금합니다! 🧐

### [feat(Calculator/hook.ts): view에 필요한 시그니처 작성](https://github.com/next-step/react-calculator/commit/eb972f588e95c7e7ee8ef3bc1edd489af7dac510) 

다음은 `UI`에 사용할 `handler`들을 `UI` 구현에 앞서 생각해보며 `hook`에 `시그니처(?)`를 작성했습니다. 제 생각엔 이 부분은 좀 과한 컨셉 같긴 합니다 😢 (ui먼저 그리고 작성해도 괜찮지 않을까 싶습니다..)

### [design(Calculator): UI 작업](https://github.com/next-step/react-calculator/commit/63742a84bb7eb99501c449ff33e6070c0f0fdb0d)

다음은 `ui`를 작업했습니다. 저는 `ui` 구현에서 중요하게 생각하는 부분이 **추상화 수준**(이라쓰고 들여쓰기 레벨)이라고 생각하고 있는데 제공되는 `css`대로 만들다보니 딱히 고려하지 않아도 됐습니다. 그리고 해당 `UI`는 간단해서 따로 `컴포넌트`로 분리하지 않아도 이해하기 어렵지 않다고 생각했습니다. 추가로 구현 [예시](https://next-step.github.io/js-calculator/)를 보면서 신기했던 부분이 `digits`와 `modifiers`의 `html 태그` 작성 순서가 보여지는 `ui`와 **반대**로 작성되어있었던데 이 부분은 접근성을 고려하여 **의도된 설계**로 이해했는데 맞을까요? 👀

### [feat(Calculator): 로직 작성 및 이벤트 함수 바인딩](https://github.com/next-step/react-calculator/commit/9883693d690c3269a5047372352744916c8d3254)

다음은 UI에 필요한 로직을 `hook`에 작성했습니다. `hook`에 작성한 이유는 상태와 연관이 되어 있기 때문입니다. 예시로 상태와 관련 없는 순수함수는 `utils` 폴더 하위에 작성했습니다.
`ui`에 필요한 상태를 각각 `useState`에 하나씩 작성 했는데 로직을 수정하면서 특정 `state`를 `set`하는걸 깜빡해 에러가 나는 경우가 왕왕 있었습니다. 항상 고민되는 부분이 연관되어 있는 값을 `useState`에 객체로 만들어 한번에 관리할지 각각 `state`로 따로 만들어 관리할지가 고민되는것 같습니다.(현재까지는 관성을 따라 각각 만드는경우가 더 많은것 같습니다.) 혹은 좀 더 나아가 `useReducer`로 변경에 대한 `actions`들을 구체적으로 다루는게 좋을까요?

로직을 작성하면서 제가 생각하지 못한 **엣지 케이스**가 많았습니다. 그려면서 함수에 If 문이 추가되어 가독성이 현저히 떨어지는것(난독화 성공)을 느꼇습니다. 😭 현업에서도 QA에서 나온 버그를 수정하면서도 이런 경험이 있었는데 **조건을 추가하는 상황속에도 가독성을 유지**할 수 있는 좋은 방법이 있을까요? 혹은 설계가 잘못됐을수도 있을것 같습니다 ㅜ



## 더 해보면 좋을 것
- 엣지 케이스에 대응하면서 기존 작성한 코드가 정상 동작하는지에 대해 확신이 없었습니다. (열심히 손으로 테스트 해볼수 밖에 없었습니다.) 이러한 점에서 테스트 코드가 있었다면 도움이 되었을것 같습니다. 차차 적용해볼 수 있도록 하겠습니다!

SaeWooKKang committed:
chore: initial commit

SaeWooKKang committed:
chore: set eslint, prettier

SaeWooKKang committed:
Merge remote-tracking branch 'react-boiler-project/main'

SaeWooKKang committed:
feat(Calculator/interface.ts): interface 작성

SaeWooKKang committed:
feat(Calculator.ts): interface 구현

SaeWooKKang committed:
feat(Calculator/hook.ts): view에 필요한 시그니처 작성

SaeWooKKang committed:
design(Calculator): UI 작업

SaeWooKKang committed:
feat(Calculator): 로직 작성 및 이벤트 함수 바인딩

SaeWooKKang committed:
refactor(utils/isNil): parameter 타입 수정

- 이전 any 작성시 eslint rule error 발생

SaeWooKKang committed:
chore(.eslintrc.cjs): 중복되어 충돌나는 룰 off 시킴

SaeWooKKang committed:
chore(.gitignore): add EOL

SaeWooKKang committed:
feat(Calculator/hook.ts): infinity값 오류 텍스트 노출 되도록 수정

greenblues1190 commented on src/components/Calculator/hook.ts:
```diff
@@ -0,0 +1,150 @@
+  const [operator, setOperator] = useState<Operator | null>(null)
+  const [showOutput, setShowOutput] = useState(false)
+
+  const calculator = new Calculator()
```
상태가 업데이트되는 매순간 Calculator 인스턴스가 새로 생성되는 것을 방지하기 위해 ref나 memo로 감싸주세요.

SaeWooKKang replied:
음.. 이 부분은 저는 성능상 문제가 안될 것(?) 같아서 처리를 안 했습니다. 아무래도 ref를 사용하는것이 좋을까요? 

아니면 상태가 변경됨에 따른 인스턴스 생성을 방지하기 위해 훅 밖 파일 상단에 선언하는 방식은 문제가 될 여지가 있을까요?

ex) 
``` js
const calculator = new Calculator()

export const useCalculator = () => {
  ...
}
```

greenblues1190 replied:
> 음.. 이 부분은 저는 성능상 문제가 안될 것(?) 같아서 처리를 안 했습니다. 아무래도 ref를 사용하는것이 좋을까요?

말씀하신대로 지금은 성능 상으로도 동작 부분이라도 문제가 되지 않아보이네요. 만약 Calculator에 내부 상태가 있다면 매번 재생성되며 상태가 초기화되는 문제가 생길 것 같아요.

> 상태가 변경됨에 따른 인스턴스 생성을 방지하기 위해 훅 밖 파일 상단에 선언하는 방식은 문제가 될 여지가 있을까요?

훅에 하나의 인스턴스가 결합되어 싱글톤이 강제되겠지만 현시점에서 문제는 없어보입니다. 확장가능성이 중요한 라이브러리들의 경우에는 Context API를 사용해 Provider로 인스턴스를 주입받는 패턴을 주로 사용합니다.

greenblues1190 commented on src/modules/Calculator/Calculator.ts:
```diff
@@ -0,0 +1,19 @@
+    return a * b
+  }
+
+  division(a: number, b: number): number {
+    return a / b
+  }
```
0으로 나누는 것에 대한 예외처리가 필요합니다.

SaeWooKKang replied:
0으로 나눴을 때 Infinity 값을 처리하라는 말씀이 맞을까요? 그렇다면 여기서 UI에서 처리할 '오류'라는 텍스트를 작성하게 되는건가요?

greenblues1190 replied:
0/0을 할 때 NaN이 발생하는 것을 우려했는데 잘 동작하네요!

greenblues1190 commented on src/modules/Calculator/Calculator.ts:
```diff
@@ -0,0 +1,19 @@
+
+  multiplication(a: number, b: number): number {
+    return a * b
+  }
```
add와 minus와 네이밍을 맞추어 동사로 바꿔주세요.

SaeWooKKang replied:
ㅎㅎ 반영했습니다. 꼼꼼하게 확인 해주셔서 감사합니다!

[chore: use verb (multiplication -> multiply)](https://github.com/next-step/react-calculator/pull/55/commits/f00bc317d6f42eeaf4860f0f7a2b56c3fa9a3e30)

greenblues1190 commented on src/components/Calculator/hook.ts:
```diff
@@ -0,0 +1,150 @@
+  const getResult = () => {
+    if (operator === '+') {
+      return calculator.add(number1!, number2!)
+    }
```
number1, number2에 대해 non-null assertion을 하신 이유가 있나요? 다른 개발자들이 사용한다고 가정했을 때 코드레벨에서 안전성을 강제하려면 타입 가드를 작성해야 할 것으로 보입니다.

SaeWooKKang replied:
number1, number2가 있을때 getResult 함수가 호출되도록 작성해서 해당 함수에서는 항상 값들이 있다고 가정했습니다.  말씀하신대로 타입가드 작성하는 것이 좋을것 같습니다!
 
[refactor(hook.ts): use type guard (delete non-null assertion)](https://github.com/next-step/react-calculator/pull/55/commits/2863e6e0ff56c78a74127cad76284d186d0fc98b)

greenblues1190 commented on src/components/Calculator/hook.ts:
```diff
@@ -0,0 +1,150 @@
+    }, '')
+  }
+
+  const parseResultWindow = () => {
+    if (showOutput) {
+      const res = getResult()
+
+      return Number.isFinite(res) ? res : '오류'
+    }
+
+    return getUserInputs()
+  }
```
res가 Inifity인 경우, NaN인 경우, 등 앞으로 예외 상황이 늘어날 경우 예외처리를 용이하게 하려면 에러 상태를 따로 관리하면 좋을 것 같습니다.

SaeWooKKang replied:
말씀하신것처럼 요구 사항이 늘어남에 따라 에러 상태를 따로 관리하면 좋을것 같은데.. 감이 안잡혀서 그런데 어떤식으로 관리하면 좋을지 힌트를 얻을 수 있을까요?

greenblues1190 replied:
미리 정의된 에러 코드를 error 리액트 상태에 업데이트하여 관리하는 예시입니다.

```tsx
const useCalculator = () => {
  const [expression, setExpression] = useState();
  const [result, setResult] = useState();
  const [error, setError] = useState();
  const [status, setStatus] = useState<'input', 'result', 'error'>();
  
  return {
    result,
    error,
    status,
    ...
  }
};

const App = () => {
  const { result, error, status } = useCalculator();
  
  useEffect(() => {
    if (error === ...) {
     /**
      * 에러 코드에 대한 뷰로직 작성 (예: 얼럿 노출)
      */
    }
  }, [])
  
  return (
    <h1 id="total">
      {status === 'input' && expression}
      {status === 'result' && result}
      {
        /**
         * 에러 코드에 대한 메세지 표시
         */
        status === 'error' && ERROR_MESSAGE[error]
      }
    </h1>
  )
};
```

greenblues1190 commented on src/components/Calculator/hook.ts:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    setNumber2(Number(fullValue))
+  }
```
검증 및 상태 업데이트에 대한 분기가 복잡해 요구사항이 늘어날 경우를 위해 책임을 분리할 필요가 있어 보여요.

개인적으로 폼 이벤트에서 검증 로직은 값과 테스트 함수만을 받아 validator 객체로 상단에 작성하고(zod 등의 런타임 검증 라이브러리를 사용해도 좋습니다) try/catch를 사용해 실패, 성공 케이스에 대해서 블록을 구분하는 것을 선호해요. 또 해당 부분은 리액트 의존성도 없어 Calculator 모듈 내부로 추상화도 가능해보입니다.

```ts
const onClickNumber = (value: number) => () => {
  try {
    // 검증 및 에러 throw
    validator.test(value, context);
    
    // 입력 처리
  } catch(e) {
    // 에러 처리
  }
}
```

이번 미션에서는 과하다고 생각하는 방법으로 `number1`, `number2`, `operator`라는 값으로 각각 관리하지 않는 방법으로는, 전체 수식 문자열을 상태로 관리하고 수식 문자열에 `change` event가 발생할 때 1) 정규표현식으로 검증하거나, 2) 토크나이저, 렉서를 사용(라이브러리가 좋겠죠?)하여 트리를 구성한 다음 트리를 순회하며 값을 계산하는 방식으로 계산기를 구현한다면 복잡한 요구사항  대응이나 파서에서 유효성 검증을 좀더 용이하게 할 수 있을 거라 생각이 듭니다!

SaeWooKKang replied:
>  2) 토크나이저, 렉서를 사용 

이 부분은 생소하네요 찾아보겠습니다!

리뷰 반영해보았습니다! 

[refactor(hook.ts): onlickNumber 핸들러 성공 실패 케이스 분](https://github.com/next-step/react-calculator/pull/55/commits/19673afd545e15a3d6f1dbe03dbc7256fdbb3020)

greenblues1190 reviewed (changes requested):
안녕하세요 재영님 리뷰로 함께 하게 되어 반갑습니다!

구현하실 때 고려하셨던 점들을 PR 본문에 자세히 적어주셔서 리뷰하는데 많은 도움이 되었어요! 작성하신 부분들에 대한 제 생각 말씀드릴게요.

> 계산기 미션에서 계산 로직은 ui 혹은 더 나아가 라이브러리가 달라져도 재사용할 수 있다고 생각했습니다. 그래서 해당 로직을 분리(moduels/Calculator)하여 작성했습니다.

계산기 로직을 모듈화하여 리액트 커스텀 훅과 계층적으로 잘 분리하셨어요!

> 다른 연산들과 달리 나누기 연산에서 '함수 파라미터로 자릿수를 받을수 있도록 구현하는게 좋을까?' 에 대해서 고민이 됐습니다. 제 생각에는 '사용하는 쪽에서 결과값을 가공해서 보여주면 되지 않을까?' 싶어서 단순히 연산만 처리하도록 작성했습니다. 우디님이라면 어떻게 처리 하실지 궁금합니다! 🧐

클래스는 인스턴스의 상태를 관리하거나 객체를 추상화하기 위해 사용하는데 간단한 사칙연산 유틸 함수의 경우 클래스로 구현할 필요성은 적게 느껴집니다. 하지만 함수들을 한가지 책임을 가진 순수함수로 만드는 것은 좋은 접근 방식이라고 생각해요.

> 다음은 UI에 사용할 handler들을 UI 구현에 앞서 생각해보며 hook에 시그니처(?)를 작성했습니다. 제 생각엔 이 부분은 좀 과한 컨셉 같긴 합니다 😢 (ui먼저 그리고 작성해도 괜찮지 않을까 싶습니다..)

설계를 먼저하는 것도 좋은 접근 방식 같습니다!

> 다음은 ui를 작업했습니다. 저는 ui 구현에서 중요하게 생각하는 부분이 추상화 수준(이라쓰고 들여쓰기 레벨)이라고 생각하고 있는데 제공되는 css대로 만들다보니 딱히 고려하지 않아도 됐습니다. 그리고 해당 UI는 간단해서 따로 컴포넌트로 분리하지 않아도 이해하기 어렵지 않다고 생각했습니다. 추가로 구현 [예시](https://next-step.github.io/js-calculator/)를 보면서 신기했던 부분이 digits와 modifiers의 html 태그 작성 순서가 보여지는 ui와 반대로 작성되어있었던데 이 부분은 접근성을 고려하여 의도된 설계로 이해했는데 맞을까요? 👀

css 부분은 의도와 다르게 잘못 작성되었다고 생각해요. 코드로만 읽었을 때는 row-reverse로 되어있어 [0, ..., 9] 배열을 렌더링하면 넘버패드 순서로 보일 줄 알았는데 이상하게 렌더링되네요.

> ui에 필요한 상태를 각각 useState에 하나씩 작성 했는데 로직을 수정하면서 특정 state를 set하는걸 깜빡해 에러가 나는 경우가 왕왕 있었습니다. 항상 고민되는 부분이 연관되어 있는 값을 useState에 객체로 만들어 한번에 관리할지 각각 state로 따로 만들어 관리할지가 고민되는것 같습니다.(현재까지는 관성을 따라 각각 만드는경우가 더 많은것 같습니다.)

휴먼에러를 줄이기 위해서는 당연히 상태를 적게 관리하고 한가지 상태에서 여러가지 값을 파생하여 원천 값이 변경되면 나머지 파생 변수들이 자동 갱신되게끔 하는 것이 좋습니다.

> 혹은 좀 더 나아가 useReducer로 변경에 대한 actions들을 구체적으로 다루는게 좋을까요?

요구사항의 복잡도에 따라 다를 것 같아요. 유지보수를 위한 가독성을 고려해야겠죠?

> 로직을 작성하면서 제가 생각하지 못한 엣지 케이스가 많았습니다. 그려면서 함수에 If 문이 추가되어 가독성이 현저히 떨어지는것(난독화 성공)을 느꼇습니다. 😭 현업에서도 QA에서 나온 버그를 수정하면서도 이런 경험이 있었는데 조건을 추가하는 상황속에도 가독성을 유지할 수 있는 좋은 방법이 있을까요? 혹은 설계가 잘못됐을수도 있을것 같습니다 ㅜ

이부분은 hook.ts 34~83번 라인에 피드백 남겼습니다.

> 엣지 케이스에 대응하면서 기존 작성한 코드가 정상 동작하는지에 대해 확신이 없었습니다. (열심히 손으로 테스트 해볼수 밖에 없었습니다.) 이러한 점에서 테스트 코드가 있었다면 도움이 되었을것 같습니다. 차차 적용해볼 수 있도록 하겠습니다!

TDD 적용해보시는 것도 좋은 학습이 되실거에요!

SaeWooKKang committed:
chore: use verb (multiplication -> multiply)

SaeWooKKang committed:
refactor(hook.ts): use type guard (delete non-null assertion)

SaeWooKKang committed:
refactor(hook.ts): onlickNumber 핸들러 성공 실패 케이스 분

- 현재 스텝을 구하는 함수(getCurrentStep), 현재 스텝에 맞는 값을 구하는 함수(getValueForCurrentStep), 유효성 검사를 하는 validator 함수로 역할 분리
